sw/device/lib/dif/BUILD

```diff
@@ -0,0 +1,34 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+cc_library(
+  name = "clkmgr",
+  srcs = [
+      ":dif_clkmgr.c",
+      ":dif_clkmgr.h",
+      ":dif_warn_unused_result.h",
+      ],
+  deps = [
+      "//sw/device/lib/base:mmio",
```

Comment:
@cfrantz this target should definitely depend on ...base:mmio as it does here, but I'm running into trouble where I need mmio.h to be built with -DMOCK_MMIO when it's used in a test. The issue appears to be discussed here: https://stackoverflow.com/questions/53950443/build-unit-test-for-c-with-mocked-dependencies-using-bazel#

Reply:
It appears the DIFs may have been in a less developed state when I built my first bazel experiment.

On thing that I did differently was to separate the interface and link-target for base:mmio (see https://github.com/cfrantz/opentitan/blob/bazel/sw/device/lib/base/BUILD).  I'm not certain that will solve the problem you're pointing to here, as we want "mmio.h" to be preprocessed with `MOCK_MMIO` defined.

One way to get there is to employ the strategy used in the meson rules, which is to include the unit under test into the `srcs` of the test rule rather than it's target into `deps`:

```
cc_library(
  name = "clkmgr",
  srcs = ["dif_clkmgr.c"],
  deps = [ ... ],
)

cc_test(
  name = "clkmgr_unittest",
  srcs = [
    "dif_clkmgr_unittest.cc",
    "dif_clkmgr.c",
    ...
  ],
  deps = [ ... ],
  defines = ["MOCK_MMIO=1"],
)   
```

Reply:
That's significantly more elegant than the workaround I had in mind.
```
cc_test(
  name = "gpio_unittest",
  srcs = [
      "dif_gpio_unittest.cc",
      "dif_gpio.c",
      "dif_gpio.h",
      "dif_warn_unused_result.h",
      "//sw/device/lib/base:mmio.c",
      "//sw/device/lib/base:mmio.h",
      ],
  deps = [
      "//hw/ip/gpio/data:gpio_regs",
      "@googletest//:gtest_main",
      "//sw/device/lib/base:mmio",
      "//sw/device/lib/base/testing:mock_mmio",
      ],
  defines = [
      "MOCK_MMIO=1",
      ],
  )
```
appears to work.

Reply:
You shouldn't need to include "//sw/device/lib/base:mmio.[ch]" into the srcs of the cc_test rule.

In my fork's `bazel` branch, this is handled by `//sw/device/lib/base:base` providing the interface to mmio (ie: `mmio.h`) but not the binary resource.  The binary resource for `mmio` is provided by `//sw/device/lib/base:mmio` and this target is only needed as a final link-time decision for riscv32 cc_binaries.  Similarly, for tests, `//sw/device/lib/base/testing:testing` provides the binary resource for the `mock_mmio` library and satisfies the mmio link-time requirements, thus making it suitable for tests (you can see that `mock_mmio` actually includes the sources for `mmio`, but with `MOCK_MMIO` set, thus fulfilling all the symbol requirements).

The last bit of a sticky point is that the `cc_test` needs define `MOCK_MMIO` so when the srcs `#include "mmio.h"` they have the right preprocessor directives defined (which you already have).

To me, I think that separating interfaces from link-time deps in `//sw/device/lib/base` gives the right amount of flexibility in listing the deps for libraries and tests.  (and `memory.c` and `bitfield.c` were left in the `:base` library target because they don't have any conditional compilation and if they happen to get linked with something that doesn't use them, then the linker/librarian will simply omit them from the target).

Once we start adding support for emitting riscv32 binaries, we can even constrain `mmio` to only riscv32 targets so that it would be impossible to accidentally link it into a test meant to run on x86.